geohisto/intercommunalities.py

```diff
@@ -0,0 +1,287 @@
+    - generate intermediate states (ie. town changes during the year)
+    - recompute population from `pmun` instead of the current `ptot`
+'''
+
+
+RE_IN_PRENTHESIS = re.compile(r'(?:\(|\[)(.*?)(?:\)|\])')
+RE_WITH_PRENTHESIS = re.compile(r'((?:\(|\[).*?(?:\)|\]))')
+
+NAME_RULES = {
+    'CC': 'Communauté de Communes',
+    'CA': 'Communauté d\'Agglomération',
+    'CU': 'Communauté Urbaine',
+    'CV': 'Communauté de Villes',
+    'DISTRICT': 'District',
+    'METRO': 'Métropole',
+    'METRO69': '',
+    'MET69': '',
+    'SAN': 'Syndicat d\'Agglomération Nouvelle',
+}
+
+# Order matters
+DISTRICT_PREFIXES = (
+    ('D u ', 'DISTRICT Urbain '),
+    ('D U ', 'DISTRICT Urbain '),
+    ('D ', 'DISTRICT '),
+    ('Di ', 'DISTRICT '),
+)
+
+TYPOS = {
+    'C C ': 'CC ',
+    'C c': 'CC ',
+    'C A ': 'CA ',
+    'C a': 'CA',
+    'C U ': 'CU ',
+    'C V ': 'CV ',
+    'S A N ': 'SAN ',
+    'C Agglo ': 'Communauté d\'Agglomération ',
+}
+
+COMMON_PREFIXES = (
+    'rural', 'rurale', 'ruraux', 'rurales',
+    'urbain', 'urbaine', 'urbains', 'urbaines'
+)
+
+PREPOSITIONS = ('du ', 'de ', 'de la ', 'd\'', 'des ')
+ALWAYS_LOWER = (
+    'du', 'de', 'des', 'sur', 'sous', 'en', 'la', 'le', 'les', 'aux', 'au', 'à'
+)
+WITH_APOSTROPHE = ('d\'', 'l\'')
+
+PRE_REPLACEMENTS = {
+    '': '\'',
+    '': 'œ',
+    '.': ' ',
+    ':': '',
+    '"': '',
+}
+
+POST_REPLACEMENTS = {
+    'Reg ': 'Région ',
+    'Region ': 'Région ',
+    'Cant ': 'Canton ',
+    'Breon': 'Bréon',
+    'Saone': 'Saône',
+    'Vallee': 'Vallée',
+    'D Or': 'd\'Or',
+    'Indre - Brenne': 'Indre-Brenne',
+    'Saône-Chalaronne': 'Saône Chalaronne',
+}
+
+
+def should_replace_kind(name, kind):
+    return name.startswith(tuple(
+        ' '.join((kind, prep)) for prep in PREPOSITIONS
+    )) or name.lower().startswith(tuple(
+        ' '.join((kind.lower(), prefix, prep)) for prep in PREPOSITIONS
+        for prefix in COMMON_PREFIXES
+    ))
+
+
+def fix_typos(name):
+    for typo, fix in TYPOS.items():
+        if typo in name:
+            name = name.replace(typo, fix)
+    return name
+
+
+def fix_word_casing(word):
+    if any(word.lower().startswith(prefix) for prefix in WITH_APOSTROPHE):
+        prefix, suffix = word.split('\'', 1)
+        return '\''.join((prefix.lower(), suffix.capitalize()))
+    elif word.lower() in ALWAYS_LOWER:
+        return word.lower()
+    elif word in NAME_RULES:
+        return word.upper()
+    else:
+        return word.capitalize()
+
+
+def fix_casing(name):
+    return ' '.join(fix_word_casing(word) for word in name.split(' '))
+
+
+def extract_name(line):
+    """Extract and normalize an intercommunality name"""
+    name = line['nom'].strip('\' ')
+    for char, replacement in PRE_REPLACEMENTS.items():
+        name = name.replace(char, replacement)
+    kind = line['nature']
+    name = fix_typos(name)
+    name = ' '.join(name.split())  # replace multiple spaces
+    if name.isupper():
```

Comment:
Good idea, done, but I don't know why it seems slower (test execution take `0.05 seconds` instead of `0.04 seconds` before the change)